### PR TITLE
Reorganise docs

### PR DESCRIPTION
**Pull request by gouttegd**
Resolves [#330]

- [x] `docs/` have been added/updated if necessary
- [x] `make test` has been run locally
- [ ] tests have been added/updated (not applicable)
- [ ] [CHANGELOG.md](https://github.com/mapping-commons/sssom/blob/master/CHANGELOG.md) has been updated (not applicable)

This PR reorganises the documentation, especially the specification part, as suggested in #330.

More precisely:

* Several bits of informations that were scattered throughout the website are now regrouped on the index page (e.g. “contacts” and “credits”).
* The home page gets a “SSSOM at a glance” section to give an immediate glimpse of what the standard is about.
* The specification is broken down as follows:
  * general introduction
  * specification of the data model
    * introduction and notes complementary to the LinkML-generated documentation
    * LinkML-generated documentation
  * specification of the serialisation formats
    * SSSOM/TSV format (completely rewritten)
    * OWL/RDF format (taken from the pre-existing `spec.md` document)
    * JSON format (currently merely a placeholder, since the JSON format is unspecified #321)

The “resources for users” section is left untouched for now. The urgent part was reorganising the _specification_, so that we can start enriching it to make it ready for 1.0.

### CODE REVIEW

**Review by matentzn — COMMENTED** [review]
Absolutely fantastic work @gouttegd 

I think our main disagreement right now is the condensation requirement. I love that the concept of condensation was introduced, I love the name, and I love the specification - I just disagree with mandating it in "strict" mode at the moment, or encouraging it in "non-strict" mode. (Unless, of course, I have already agreed to it elsewhere and just forgot about it)

I hope we did

File: src/docs/introduction.md
Comment (gouttegd) [reply src/docs/introduction.md:5]: Please reserve all comments about the `introduction.md` file for another issue. This file comes straight from the introduction section of the old `spec.md` file, I didn't change anything in there. All issues you may find were already present.

As I've mentioning in the associated commit message, I believe this entire introduction should be rewritten anyway.

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md:37]: Given that we are using YAML, we are bound to face _a lot_ of variability anyway.

In fact that’s one of the main reasons for defining a “canonical” variant -- to reduce the variability.

But people are going to expect (are already expecting) to be able to do in the metadata block whatever YAML allows them to do, and I believe we should not stop them from doing so. Otherwise what would be the point of using YAML? We might as well define our own format to store the metadata block, without any of the variability of YAML.

YAML variability shouldn't be a problem for most implementations since they are most likely going to delegate parsing the metadata block to a full-blown YAML parser anyway (that's what both SSSOM-Py and SSSOM-Java are doing). In fact, trying to allow only a subset of what YAML allows is likely going to be _more difficult_ for implementations, because then they would have to override their YAML parser. I see no reason for doing that.

In the case at hand, the problem is the opposite: we are allowing something that YAML does not (representing a one-item list as a scalar value). But that form is widespread among SSSOM mapping sets. In fact, until [I corrected it a few months ago](https://github.com/mapping-commons/sssom/pull/362), this form was even found in the examples given in the spec! SSSOM-Py has always accepted that form when reading, and is still generating it when writing. I don't like that (it forced to do a lot of special-casing in SSSOM-Java’s parser), but we cannot outlaw that form now without breaking a lot of existing sets.

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md:112]: I outlined the various options in [our discussion here](https://github.com/mapping-commons/sssom/issues/305#issuecomment-1653297621). The above corresponds to option 3.

You were clearly against option 2 and I explained in [a following comment](https://github.com/mapping-commons/sssom/issues/305#issuecomment-1656810155) the problems with option 1, concluding that option 3 was “the safest and least confusing option” -- something [you agreed with](https://github.com/mapping-commons/sssom/issues/305#issuecomment-1657186217).

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md:116]: It may be “new” in that we did not discuss it in #305, but it does not come from nowhere. This is the logical consequence of allowing propagation, since it is the opposite operation. You would have nothing to propagate if there wasn’t any condensation to begin with.

> In most cases, people wont want to condensate

Why? That’s an internal detail of the format. Why would people care about how the metadata are stored, as long as the metadata _are_ stored and can be retrieved whenever needed?

That would be like saying that people using ZIP files won’t want their data to be compressed with the DEFLATE algorithm. Why would they care?

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md:178]: So either propagation and condensation should both be recommended, or they should both be optional. But if they are optional, then I’d question why we even bother having “propagatable slots” at the mapping set level.

If propagation is optional, we can expect that many implementations will not support it (they won’t have to -- it’s optional). Then it means that people (even those crafting their mapping sets manually) will be wary of using propagatable slots, because they will not be sure whether they will be interpreted (propagated) correctly.

So no one will ever use propagatable slots, and the entire feature will exist for nothing. Then, again, why do we bother having this feature? We might as well remove all “propagatable slots” from the `MappingSet` class.

File: src/docs/index.md
Comment (gouttegd) [reply src/docs/index.md:84]: Hum, no _strong_ objection, but I don’t like having _anything_ in the documentation that would be merely a pointer to GitHub. The documentation should be, as much as possible, “standalone”.

In fact, if the list of contributors is merely a pointer to a GitHub team, this creates a problem, because if a contributor leaves the team at so point, then he or she will no longer appear in the documentation, as if they had never been a contributor in the first place. I believe this is wrong. If someone has contributed to the doc/the spec, their name should stay in the list even after they leave the team. 

~~(Unless they leave the team to work on a competing project, in which case their name should fall into oblivion for eternity.)~~

File: src/docs/index.md
Comment (gouttegd) [reply src/docs/index.md:38]: No objection, but:

1) I will use a simplified version of that example (with less metadata). The point in the “at a glance” section is to have a _basic_ example to give the reader an idea of what the format looks like – it does not have to showcase the various metadata slots.

2) I will leave aside the mapping that uses the `sssom:NoMapping` special identifier, which, unless I have missed something, is not specified anywhere (and the IRI does not resolve to anything). I know that special identifier has been discussed in #28, but (a) in the end the agreement was to use `sssom:NoTermFound` rather than `sssom:NoMapping`, and (b) even after the agreement was reached at the end of the discussion, the special identifier did _not_ find its way into the spec.

If that special identifier is still a desirable feature (a question that is worth asking in my opinion, since the aforementioned discussion is almost three years old and seemingly nobody has complained that the special identifier was still missing), then I’d suggest that it be added to the `spec-model.md` document after this PR has been merged.

3) Less important, but that example is invalid as it is using `orcid:…` and `wikidata:…` CURIEs without declaring the `orcid` and `wikidata` prefix names. We should _really_ pay attention and make sure that we are giving examples that would _not_ be rejected by our own tools!

```
Invalid SSSOM data in file foodieinc-food.sssom.tsv: Some prefixes are undeclared: orcid, wikidata
```

Comment (gouttegd) [reply src/docs/index.md:38]: Also, that comment at the mapping set level is weird:

```
#comment: We could map to FOODON:00004187 instead which more specifically refers to
#  "raw" Pink apples. Decided against to be consistent with other mapping choices.
```

Looks like a duplicate of the mapping-level comment in the `KF_FOOD:0003 -- skos:exactMatch -- FOOD:00004186` mapping, and almost certainly the result of some kind of error.

File: src/docs/index.md
Comment (gouttegd) [reply src/docs/index.md:89]: Of note, if we do keep the list here (and update it of course):

I just realised that the ORCID organisation is one of those that have the gall to [forbid some kind of links](https://info.orcid.org/brand-guidelines/#h-orcid-logos-and-icons) to their websites:

> We do not currently allow references to authorized ORCID iDs without using the iD icon.

So a link like “[Damien Goutte-Gattat](https://orcid.org/0000-0002-6095-8718)” is _not allowed_, since the ORCID iD icon is missing. Instead it should be something like

* [ https://orcid.org/0000-0002-6095-8718](https://orcid.org/0000-0002-6095-8718) (icon + full URL)
* [ 0000-0002-6095-8718](https://orcid.org/0000-0002-6095-8718) (icon + ORCID iD)
* [Damien Goutte-Gattat ](https://orcid.org/0000-0002-6095-8718) (name + icon)

(It will never not amaze me that some people believe they can dictate whether and how other people can create links… But I am more used to see that from big old-fashioned business brands that know nothing about how the web works, it’s disappointing to see that from an organisation like ORCID that is both not-for-profit and supposedly tech-savvy. Oh well…)

Your call whether you want to follow their rules or not. Personally, I am inclined to disregard their rules entirely, and to create links on our own website in whatever way we want, whether they _allow_ it or not.

File: src/docs/index.md
Comment (matentzn) [reply src/docs/index.md]: Looks great! Thank you! 

> Also, that comment at the mapping set level is weird:

Looks like an oversight, agree.

File: src/docs/introduction.md
Comment (matentzn) [reply src/docs/introduction.md:5]: Ok!

File: src/docs/spec-formats-tsv.md
Comment (matentzn) [reply src/docs/spec-formats-tsv.md:112]: Yeah, I still believe this is ok. Sorry for noise, resolving.

File: src/docs/index.md
Comment (matentzn) [reply src/docs/index.md:89]: Shall we leave updating this list to a separate issue? If so, you could turn your comment into an issue and assign to me, I will update the contributor lists and think about the ID icon.

File: src/docs/spec-formats-json.md
Comment (matentzn) [reply src/docs/spec-formats-json.md:5]: Ok!

File: src/docs/spec-formats-tsv.md
Comment (matentzn) [reply src/docs/spec-formats-tsv.md:37]: Well, you are arguing for a less draconian spec so I wont stand in the way! I am fine with this, if need be, we can tighten the screws in a future version 2.0.

File: src/docs/spec-formats-tsv.md
Comment (matentzn) [reply src/docs/spec-formats-tsv.md:191]: No strong feelings, if you thought about it, I am fine with following your nose.

File: src/docs/spec-formats-tsv.md
Comment (matentzn) [reply src/docs/spec-formats-tsv.md]: I dont know if I agree with the analogy here. I am not so concerned about SHOULD vs MAY, but I am concerned that tools will take the SHOULD as a permission to build serialisers that do not even _permit_ un-normalised / un-condensed outputs.. I want to be clear that if a user _wants_ to serialise an uncondensed version, they should be able to.

File: src/docs/index.md
Comment (gouttegd) [reply src/docs/index.md:89]: Fine with me.

> I will update the contributor lists.

If we agree that, in the docs/spec, the list of contributors is understood as the list of “all the persons who have ever contributed, regardless of whether they are still contributors _now_”, then updating the list should only be a matter of _adding_ the persons who started contributing after that list was created, right? Basically the list is append-only – no one should ever be removed from it. Do we agree on that?

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md]: But even so: what would be the problem if you have a tool that only generate condensed output?

First, for those who craft their sets “manually”, they would still be able to write them in whatever way they want. Readers that support reading condensed files (that is, readers that supports propagation, which should be most readers since propagation is a SHOULD) will always support reading uncondensed files.

And for those who create sets using a tool that only generate condensed output: again, what is the problem? As long as readers SHOULD support propagation (which they really should, otherwise as I said further below, the whole idea of having propagatable slots becomes useless), condensed values will always be propagated back upon reading. Users won’t even need to know that the file has been written in condensed form.

Comment (gouttegd) [reply src/docs/spec-formats-tsv.md]: > I am concerned that tools will take the SHOULD as a permission to build serialisers that do not even permit un-normalised / un-condensed outputs..

I could understand if the concern was that implementers take the SHOULD as a permission to build _readers_ that do not support reading anything other than canonical and condensed output (“since writers SHOULD write condensed output, we should encounter very few or even zero files that are non-condensed, so we don’t need to bother supporting them”). That would be bad indeed.

But there is no risk of that.

First, it is impossible to support reading condensed files _only_. A reader that can read condensed files (that is: a reader that can perform propagation, to “un-condense” the slots that were condensed) can necessarily also read non-condensed files.

Second, regarding canonicalisation: the proposed spec explicitly forbids (MUST NOT) a reader from rejecting a file solely because it is not in the canonical format.

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md:37]: One bit of variability that we could consider forbid, though, is the use of “advanced” YAML features such as tags, anchors and references. They may be nice, but I think they would be antithetical to SSSOM being a _Simple_ standard.

That is, how about explicitly forbidding things like

* Type tags

```yaml
comment: !!str This is explicitly typed as a string
```

Explicit typing should never be needed anyway since we have the LinkML schema to always tell us what type to expect for any slot.

* Non-type tags

```yaml
comment: !mytag This is a tagged comment
```

Such tags are currently not supported by SSSOM-Py, and I don’t think we’d lose much in forbidding their use entirely.

* Anchors and references

```yaml
creator_id: &the_creators
  - ORCID:1111-2222-3333-4444
author_id: *the_creators
```

Seemingly supported by SSSOM-Py, currently not by SSSOM-Java (because not supported by the underlying YAML library). Again, I don’t think we’d lose much in explicitly forbidding this kind of stuff.

Comment (matentzn) [reply src/docs/spec-formats-tsv.md:37]: ok, yes. i never even knew these existed! definetly exclude imo

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md]: How about:

> Writers SHOULD perform condensation before writing a set, but that behaviour MUST be deactivatable.

?

Comment (matentzn) [reply src/docs/spec-formats-tsv.md]: If you are fine with this, then great, that covers my concerns fully!

File: src/docs/spec-formats-tsv.md
Comment (matentzn) [reply src/docs/spec-formats-tsv.md:178]: Now that I get the point better, I agree with this! I read this section a bit differently the first time around (like a value judgement - your Mapping set is somehow worse when it is not in normal form), but having a clearly defined _canonical_ form is obviously very valuable for many reasons.

File: src/docs/index.md
Comment (matentzn) [reply src/docs/index.md:89]: You are right, which means that these list should, indeed, be managed separately. I just need to do a 1 time sync between the github core sssom developer list now, and then these list are forever disconnected, because, when someone chooses to leave the sssom-core team for whatever reason, this does not mean they loose editor status

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md]: Deactivating condensation in SSSOM-Java's TSVWriter is a matter of putting _one_ function call behind a conditional expression, so if it's all it takes to get you to agree that condensation SHOULD be supported, I'm a fine with that. :)

I'll also add something like

> Condensation and propagation work in concert. Therefore, an implementation that supports one operation MUST support the other.

This doesn't change the fact that both operations remain a "SHOULD" requirement.

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md:37]: > i never even knew these existed!

Pretty sure you’re not the only one. :) The complexity of YAML is regularly underestimated.

File: src/docs/spec-formats-tsv.md
Comment (gouttegd) [reply src/docs/spec-formats-tsv.md:178]: > like a value judgement - your Mapping set is somehow worse when it is not in normal form

That is absolutely not how the ”canonical format” should be understood, and if you have any suggestion as to how make that clearer in the spec, I’m listening.

Basically, the point of the canonical format is to avoid diffs like this:

```diff
-UBERON:0001931   skos:exactMatch   BTO:0006166   semapv:ManualMappingCuration   0.556
+UBERON:0001931   skos:exactMatch   BTO:0006166   semapv:ManualMappingCuration   0.556000
```

where the mapping has not actually changed, but the file is different all the same because the old file was written by an implementation that truncated floating values to 3 digits after the decimal point, while the new file was written by an implementation that truncates floating values to 6 digits.

Comment (matentzn) [reply src/docs/spec-formats-tsv.md:194]: I think this is fine. The spec does not need to provide guidelines on how to "feel about the spec". I was missing the thought "canonical format --> diff" and if I would have had that in my mind like a good engineer I would not have even stumbled across this. All good.

**Review by matentzn — APPROVED** [review]
I approve:

1. The reorganisation overall
2. The general sentiment expressed by the spec-formats sections that are new

I reserve myself the right the revisit details on the implementation side if (and only if) putting them into action will result in a violation of current common practice. I don't think there is anything, I just want to be super transparent. 

Thank you @gouttegd I am happy with this. When you are happy, what do you suggest:

1. Merge bypassing the need for a second review on the grounds that little new is added to the spec (only clarifications and interpretation of the spirit)
2. Me to find a second reviewer to fulfill the 2-reviewer requirement

I will leave the choice to you, I am ok with either.

**Comment by gouttegd** [comment]
Given that the “reorganisation of the docs” part does not actually change any _content_ (it merely puts the doc in a shape that it will make it easier to work on it), I am fine with that part not being reviewed by a second reviewer.

For the `spec-formats-tsv.md` part, however, I don’t think it is fitting that the first real formal specification of the SSSOM/TSV format has been written by the developer of only one of the two “major” implementations. I’d like a SSSOM-Py developer to have at least a cursory look at it.

I don’t foresee any problem since the new spec should be fully compatible with existing behaviours in SSSOM-Py. What the new spec does add:

* Propagation/condensation: As far as I know SSSOM-Py has currently no support for these concepts. The spec proposes them as recommended (_SHOULD_), so I’d like to know whether SSSOM-Py developers think they can implement them (alternatively, it could be done by an additional layer _on top_ of SSSOM-Py).
* “Canonical format”: I’ve tried to select canonical rules that would minimize the work needed for SSSOM-Py to generate canonical output. For example, the rule about the formatting of floating point values has been directly inspired by what SSSOM-Py already does, so no change in SSSOM-Py should be needed here.
* Backwards compatibility with SSSOM < 0.9.1: SSSOM-Py does not support that, but that is very clearly an entirely optional feature (we both agreed on that when we discussed backwards compatibility options), so that shouldn’t be a problem. SSSOM-Py developers can decide that they won’t ever do that.

**Comment by matentzn** [comment]
I feel myself responsible for the sssom py implementation, even though the majority of the work has been done by @hrshdhgd.

@hrshdhgd - feel free to review the file called `spec-formats-tsv.md` in this PR, with a specific emphasis on the points @gouttegd made in his [last comment above](https://github.com/mapping-commons/sssom/pull/368#issuecomment-2213522522). We will have to implement condensation and propagation at some point soon after, so it is in any case good if you are familiar with it. Let us know if you have any major qualms!

Thanks!

**Review by hrshdhgd — APPROVED** [review]
This is  so thorough , I love it! 

I don't think I follow the condensation and propagation concepts though. Could either of you provide examples so I understand what to implement?

**Comment by gouttegd** [comment]
@hrshdhgd 

> I don't think I follow the condensation and propagation concepts though. Could either of you provide examples so I understand what to implement?

Let’s consider the following set:

```
#curie_map:
#  COMENT: https://example.com/entities/
#  ORGENT: https://example.org/entities/
#mapping_provider: https://example.org/provider
#mapping_tool: foo mapper
subject_id    subject_label   predicate_id      object_id     object_label   mapping_justification          mapping_tool
ORGENT:0001   alice           skos:closeMatch   COMENT:0011   alpha          semapv:ManualMappingCuration   
ORGENT:0002   bob             skos:closeMatch   COMENT:0012   beta           semapv:ManualMappingCuration   bar mapper
ORGENT:0004   daphne          skos:closeMatch   COMENT:0014   delta          semapv:ManualMappingCuration   
ORGENT:0005   eve             skos:closeMatch   COMENT:0015   epsilon        semapv:ManualMappingCuration   
```

The set-level metadata contain a value for the `mapping_provider` and `mapping_tool` slots. These slots are considered “propagatable“, which means that they really apply to individual mappings, and that putting them at the level of the set is just a “shortcut” to avoid repeating the same value for all mappings.

So in this example, all mappings should be considered to have a `mapping_provider` of `https://example.org/provider`.

Propagation is the act of taking the values of propagatable slots at the set level, and filling the corresponding slots in each individual mappings.

After propagation, the above set should look like this:

```
#curie_map:
#  COMENT: https://example.com/entities/
#  ORGENT: https://example.org/entities/
#mapping_tool: foo mapper
subject_id    subject_label   predicate_id      object_id     object_label   mapping_justification          mapping_tool   mapping_provider
ORGENT:0001   alice           skos:closeMatch   COMENT:0011   alpha          semapv:ManualMappingCuration                  https://example.org/provider
ORGENT:0002   bob             skos:closeMatch   COMENT:0012   beta           semapv:ManualMappingCuration   bar mapper     https://example.org/provider
ORGENT:0004   daphne          skos:closeMatch   COMENT:0014   delta          semapv:ManualMappingCuration                  https://example.org/provider
ORGENT:0005   eve             skos:closeMatch   COMENT:0015   epsilon        semapv:ManualMappingCuration                  https://example.org/provider
```

Notice that that the set no longer have a `mapping_provider` value, and conversely that all mappings have one.

Also note that the value of the `mapping_tool` at the set level (“foo mapper”) has _not_ been propagated, even though `mapping_tool` is a propagatable slot. This is because one of the mappings already had a value for that slot (mapping #2, which has the value “bar mapper“), and propagation is only allowed when _no mappings at all_ have a value for the propagatable slot.

**Comment by gouttegd** [comment]
Condensation is the exact opposite of propagation. It’s taking the values of “propagatable slots” that are set on the mappings, and moving them (if possible, that is if they all mappings have the same value) to the level of the set instead.

For example, to condense the second example from my previous message, you would observe that all mappings have the same value for the `mapping_provider` slot, so you would set a single `mapping_provider` slot at the level of the set and remove the entire `mapping_provider` column. You would also observe that _not_ all mappings have the same value for `mapping_tool` (one mapping has the value “bar mapper“, whereas other mappings have no value), so you would not do anything special for that slot (it is not condensable).

**Comment by hrshdhgd** [comment]
That makes perfect sense! Thank you for explaining this patiently and perfectly @gouttegd ! I truly appreciate it.

**Pull request merged**
